Speed up cache check for locations fixture

esoergel opened the pull request:
## Product Description


## Technical Summary
https://dimagi.atlassian.net/browse/USH-5935
This includes a lot of clean up - definitely read by commit, should make sense that way.
Highlights:

* Remove `SYNC_ALL_LOCATIONS` feature flag
* Simplify/clarify (in my opinion) the fixture status stuff
* Add some deferred evaluation to speed up the cache check

## Feature Flag


## Safety Assurance


### Safety story
lots of automated tests

### Automated test coverage


### QA Plan


### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Remove deprecated, unused feature flag

This made restores return all locations.  It isn't in use by any active
projects.

esoergel committed:
user.locations doesn't exist

It was removed in https://github.com/dimagi/commcare-hq/pull/14791
Nothing references this anymore, and if it did, it would 500

esoergel committed:
Consolidate duplicate methods

esoergel committed:
Use IntegerChoices

esoergel committed:
Move bulk update thing onto the model itself

Doesn't really make sense for this to live with users

esoergel committed:
Move getter util onto the model

esoergel committed:
Remove a bunch of indirection

esoergel committed:
update_or_create simplifies this

esoergel committed:
Simplify

esoergel committed:
Defer queryset instantiation and exists check

This way if no locations on the domain have changed, we can just return
an empty fixture without doing any of that - should make caching faster

esoergel committed:
lint

esoergel commented:
@coderabbitai review

esoergel commented on corehq/apps/locations/fixtures.py:
```diff
@@ -63,23 +64,31 @@ def should_sync_locations(last_sync, locations_queryset, restore_state):
+    if not SQLLocation.objects.filter(domain=restore_user.domain, last_modified__gte=last_sync.date).exists:
+        return False
+    return accessor.queryset.filter(last_modified__gte=last_sync.date).values('last_modified').union(
         LocationType.objects.filter(domain=restore_user.domain,
```
good bot

millerdev reviewed (commented):
Beautiful cleanup! The commits were very nicely arranged too. Looks good, but I was tired when I did my review. I'd like the USH team to also review.

Jtang-1 reviewed (commented):
really cool to see the strategy used to improve performance - looks good to me, just the same question that Daniel has

esoergel commented on corehq/apps/locations/fixtures.py:
```diff
@@ -63,26 +64,31 @@ def should_sync_locations(last_sync, locations_queryset, restore_state):
+    # this first check is much faster - short circuit out if nothing at all changed
+    if not SQLLocation.objects.filter(domain=restore_user.domain, last_modified__gte=last_sync.date).exists():
+        return False
+    return accessor.queryset.filter(last_modified__gte=last_sync.date).values('last_modified').union(
```
Oh yeah I meant to check up on that too - pretty sure it's useless.  I'll see if tests complain

MartinRiese commented on corehq/apps/locations/fixtures.py:
```diff
@@ -247,13 +253,20 @@ def should_sync_flat_fixture(project, app):
-    if not user_location_pks:
-        return SQLLocation.objects.none()
-    return _location_queryset_helper(user.domain, user_location_pks)
+        if not user_location_pks:
```
was that just dead code before?

esoergel committed:
Move some code around to prep for refactor

esoergel committed:
Rework this as a UserLocations util

esoergel committed:
move should_sync conditions in-line

This keeps all the "should sync" logic in one place, and limits the
scope of the UserLocations instance

esoergel reviewed (commented):
Just pushed a few commits cleaning up the logic a bit more.  I tried to break it apart into logical chunks, so hopefully it's straightforward enough to follow.
